### PR TITLE
(CDAP-4569) Localize logback.xml correctly for MR and Spark.

### PR DESCRIPTION
**Pull request by chtyim**
- Create a jar with logback.xml inside.
- Localize the jar and make sure it’s the first in the classpath

### CODE REVIEW

**Comment by chtyim** [comment]
Verified it is working on cluster.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java [hunk 1]
```diff
@@ -414,6 +432,31 @@ private File saveCConf(CConfiguration cConf, File targetDir) throws IOException
   }
 
   /**
+   * Creates a jar in the given directory that contains a logback.xml loaded from the current ClassLoader.
+   *
+   * @param targetDir directory where the logback.xml should be copied to
+   * @return the {@link File} where the logback.xml jar copied to or {@code null} if "logback.xml" is not found
+   *         in the current ClassLoader.
+   */
+  @Nullable
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java]: is this duplicate code?

**Comment by anew** [comment]
So the logback.xml has to be inside a jar file? We can't just put the file itself into the classpath?

**Comment by chtyim** [comment]
Classpath need to be dir or jar file. Cannot be a file. We can localize the logback.xml to the $PWD of the container and add $PWD to classpath. But I feel safer to create a JAR and make sure it's the first in the classpath so that nothing from the FS dir listing order can mess things up.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java]: It is a bit similar to MR, but not exactly the same. We can abstract out the core piece (the finding "logback.xml" from a classloader and create a jar to a output supplier) and call them with different classloader and output supplier, which I need to find a place to host that method (maybe some util class, but we don't have something similar yet). I am open to do the refactoring.

**Comment by anew** [comment]
OK. LGTM

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java (hunk 1 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java]: At some point we will have to consolidate the common pieces of the mapreduce and spark runners. We can do it then.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT3406-5

**Pull request merged**